### PR TITLE
Add periodic updates to service-experiments updater.

### PR DESCRIPTION
**Pull request by travis79**
Use WorkManager along with a PeriodicWorkRequest to update experiments every 6 hours.

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt [hunk 1]
```diff
@@ -84,3 +123,23 @@ internal class ExperimentsUpdater(
...
+    workerParameters: WorkerParameters,
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal var updater: ExperimentsUpdater = Experiments.updater
+) : Worker(context, workerParameters) {
+
+    override fun doWork(): Result {
+        return if (updater.updateExperiments()) {
+            Result.success()
+        } else {
+            Result.failure()
```
Comment (georgf) [inline components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt]: What happens on failure?
- Do we get automatic retry from work manager?
- Or do we need a follow-up bug for retry behavior?

File: components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsUpdaterTest.kt [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    fun `ExperimentsUpdater initialize schedules ExperimentsUpdateWorker in WorkManager`() {
+        assertFalse("ExperimentsUpdateWorker should not be scheduled yet",
+            isWorkScheduled(ExperimentsUpdater.TAG))
+        experimentsUpdater.initialize(configuration)
+        assertTrue("initialize must schedule ExperimentsUpdateWorker",
+            isWorkScheduled(ExperimentsUpdater.TAG))
+    }
+
+    @Test
+    fun `ExperimentsUpdateWorker doWork returns SUCCESS`() {
```
Comment (georgf) [inline components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsUpdaterTest.kt]: What about test coverage for the failure case?

File: components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsUpdaterTest.kt [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        resetExperiments()
+    }
+
+    @Test
+    fun `ExperimentsUpdater initialize schedules ExperimentsUpdateWorker in WorkManager`() {
+        assertFalse("ExperimentsUpdateWorker should not be scheduled yet",
+            isWorkScheduled(ExperimentsUpdater.TAG))
+        experimentsUpdater.initialize(configuration)
+        assertTrue("initialize must schedule ExperimentsUpdateWorker",
+            isWorkScheduled(ExperimentsUpdater.TAG))
```
Comment (georgf) [inline components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsUpdaterTest.kt]: Should we assert the right parameters being set?

File: components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsTest.kt [hunk 4]
```diff
@@ -77,14 +81,21 @@ class ExperimentsTest {
         `when`(experimentsUpdater.getExperimentSource(configuration)).thenReturn(experimentSource)
     }
 
+    private fun mockUpdaterWorkManagerFunctionality() {
```
[the remark this answers is not in the document]
Comment (travis79) [reply components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsTest.kt]: I was nervous about running everything async, but I like the idea of adding _limited_ async testing since it feels like it would be easier to manage/control a test or two running async rather than everything.

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt [hunk 5]
```diff
@@ -25,23 +29,55 @@ internal class ExperimentsUpdater(
...
     internal fun initialize(configuration: Configuration) {
         source = getExperimentSource(configuration)
 
-        updateJob = GlobalScope.launch(IO) {
-            // Update the experiments list from the server async, without blocking
-            // the app launch.
-            updateExperiments()
-        }
+        // Schedule the periodic experiment updates
+        scheduleUpdates()
```
[the remark this answers is not in the document]
Comment (travis79) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt]: No, it should fire immediately for the first execution when the `.enqueueUniquePeriodicWork` function is called.  This does appear to be the case from my manual testing.

File: components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsTest.kt [hunk 6]
```diff
@@ -77,14 +81,21 @@ class ExperimentsTest {
...
+        // Since WorkManager doesn't support working with robolectric, need to mock running the task
+        // but doing it synchronously to avoid a lot of headaches, to guarantee we don't need to
+        // await anything.
+        `when`(experimentsUpdater.scheduleUpdates()).then {
```
[the remark this answers is not in the document]
Comment (travis79) [reply components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsTest.kt]: I believe it could, I will refactor and see.  There were several tests that didn't require it (i.e. they pass without it) which is why I split it out like this.

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt [hunk 7]
```diff
@@ -25,23 +29,55 @@ internal class ExperimentsUpdater(
     @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
     internal lateinit var source: KintoExperimentSource
 
-    @Volatile private var updateJob: Job? = null
```
[the remark this answers is not in the document]
Comment (travis79) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt]: Per the previous comments, I will add async executions back to specific test(s) in order to test this.  Using WorkManager now, it may be that it has to be faked and awaited a little differently than before.

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt [hunk 8]
```diff
@@ -84,3 +123,23 @@ internal class ExperimentsUpdater(
...
+
+/**
+ * This class represents a [Worker] object that periodically updates the experiments.
+ * @param updater used so that the updater can be injected from tests.
+ */
+internal class ExperimentsUpdaterWorker(
+    context: Context,
+    workerParameters: WorkerParameters,
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal var updater: ExperimentsUpdater = Experiments.updater
```
[the remark this answers is not in the document]
Comment (travis79) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt]: I didn't intend for it to get passed under normal circumstances, it should resort to the default of using the singleton's updater.  I added this in order to inject an `ExperimentsUpdater` from a test so that I could initialize it and manipulate it when unit testing the `ExperimentsUpdateWorker` class.

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt (hunk 1 above)
Comment (travis79) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt]: No, this is an oversight on my part.  I _intended_ to return `Result.retry()` here.

File: components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsUpdaterTest.kt (hunk 2 above)
Comment (travis79) [reply components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsUpdaterTest.kt]: It's actually hard to test that bit because of the call chain.  Let me take a stab at mocking the updater function to return a false and see if I can get coverage for that line.

File: components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsUpdaterTest.kt (hunk 3 above)
Comment (travis79) [reply components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsUpdaterTest.kt]: We absolutely can, that just seemed more like testing WorkManager than testing the scheduler, but I guess just checking that it was scheduled is sort of doing the same thing.  I'll add asserts for the default parameters just to be thorough.

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt (hunk 1 above)
Comment (travis79) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt]: Actually, what I think I was thinking is that `Result.failure()` would just give up this attempt and it would try again at the next periodic scheduled interval.

Comment (travis79) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt]: I was planning on setting up a manual test today using samples-glean to run the update on the minimum scheduled interval (15 minutes :( ) in order to make sure the periodic part was working and to measure the drift.  I'll try it with both `.failure()` and `.retry()` in order to see what the difference is, but I suspect that `retry` will retry immediately and `failure` will run at the next scheduled interval.

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (travis79) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt]: It looks like since I've added this, my manual test in samples-glean has broken because of it.  Looks like I'll have to find another way around this...

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt [hunk 9]
```diff
@@ -25,23 +29,55 @@ internal class ExperimentsUpdater(
...
 
+    /**
+     * Builds the periodic work request that will control the scheduling of the worker and provide
+     * the tag value so that the worker can be interacted with later.
+     *
+     * @return [PeriodicWorkRequest]
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun getWorkRequest(): PeriodicWorkRequest {
+        return PeriodicWorkRequest.Builder(
```
Comment (georgf) [inline components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt]: Per our conversation, this also can trigger when in the background.
This has some implications for how experiments run, we should file a follow-up bug and talk through that.

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt (hunk 1 above)
Comment (georgf) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt]: We should retry on failure and probably quickly initially.
Maybe we can find out what the normandy behavior is and align with that?

Comment (travis79) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt]: Normandy client behavior is reported to be to fail and retry in 6 hours, with the possibility that this may change in the future.  Since this is easy to iterate on, I'm going to verify that returning `Result.failure()` does what we expect (i.e. matches current Normandy client behavior) and call this good for now.

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt (hunk 9 above)
Comment (travis79) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt]: Follow up bug [here](https://bugzilla.mozilla.org/show_bug.cgi?id=1544931)

File: components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsUpdaterTest.kt (hunk 3 above)
Comment (travis79) [reply components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsUpdaterTest.kt]: added tests to cover all the parameters I could get at.

**Comment by travis79** [comment]
Review feedback has been addressed, ready for another look.

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt (hunk 5 above)
Comment (georgf) [inline components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt]: Per IRC conversation, we currently don't guarantee immediate updates on app launch.

We can change this to use .REPLACE, then we get:
- An immediate update on app launch.
- A predictable schedule (it resets to update every 6h from app launch again)

**Comment by travis79** [comment]
Final changes based on feedback around `KEEP` and `REPLACE` polices has been completed and pushed.  I also had to fix a couple of merge conflicts after the multi-branching PR landed, which are also included in the latest push.

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt [hunk 10]
```diff
@@ -88,3 +133,21 @@ internal class ExperimentsUpdater(
         private const val EXPERIMENTS_COLLECTION_NAME = "mobile-experiments"
     }
 }
+
+/**
+ * This class represents a [Worker] object that periodically updates the experiments.
+ * @param updater used so that the updater can be injected from tests.
+ */
+internal class ExperimentsUpdaterWorker(
```
Comment (georgf) [inline components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt]: Is it a typical style to have this at the bottom?

From the desktop code and other projects i'm used to having helpers & utils before the main class.
That helps with actually discovering them, i wouldn't think of looking for other classes after this files main class.

Unless there is a good reason, let's maybe move this one up to before `class ExperimentsUpdater`?

File: components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsTest.kt [hunk 11]
```diff
@@ -80,16 +86,29 @@ class ExperimentsTest {
         `when`(experiments.getExperimentsUpdater(context)).thenReturn(experimentsUpdater)
         experimentSource = source
         `when`(experimentsUpdater.getExperimentSource(configuration)).thenReturn(experimentSource)
+
+        // Since WorkManager doesn't support working with robolectric, need to mock running the task
```
Comment (georgf) [inline components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsTest.kt]: Super nit: "we need to"?

File: components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsTest.kt [hunk 12]
```diff
@@ -80,16 +86,29 @@ class ExperimentsTest {
...
+        // with a way to execute synchronously to avoid a lot of headaches, but also supporting
```
Comment (georgf) [inline components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsTest.kt]: Super nit: "... headaches. For some test cases we also ..."?
The full sentence was hard to parse for me.

File: components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsTest.kt [hunk 13]
```diff
@@ -802,15 +793,31 @@ class ExperimentsTest {
...
         file.writer().use {
             it.write("""{"experiment":[""")
         }
 
         val experimentStorage = FlatFileExperimentStorage(file)
 
-        resetExperiments(experimentSource, experimentStorage)
+        // Because we interact with experimentsUpdater in resetExperiments in order to setup mocking
+        // of the update process, we need to use a scoped down version of the reset here to pass the
+        // verify zero interactions assertion
```
Comment (georgf) [inline components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsTest.kt]: Can we use `verifyNoMoreInteractions()`?

File: components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsTest.kt [hunk 14]
```diff
@@ -99,7 +118,24 @@ class ExperimentsTest {
 
     @Test
     fun `initialize experiments before glean`() {
-        resetExperiments()
+        // Because we interact with experimentsUpdater in resetExperiments in order to setup mocking
+        // of the update process, we need to use a scoped down version of the reset here to pass the
+        // verify zero interactions assertion
```
Comment (georgf) [inline components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsTest.kt]: Can't we use `verifyNoMoreInteractions()` here instead?

File: components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsUpdaterTest.kt [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+        val workInfo = getWorkInfoByTag(ExperimentsUpdater.TAG)
+        assertNotNull("workInfo must not be null", workInfo)
+        assertEquals("workInfo id must match request id", workRequest.id, workInfo?.id)
+    }
+
+    @Test
+    fun `ExperimentsUpdateWorker doWork returns SUCCESS`() {
+        experiments.initialize(context, configuration)
+        experimentsUpdater.initialize(configuration)
+        // Replace the updater in the singleton with the one built for the test
```
Comment (georgf) [inline components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsUpdaterTest.kt]: Super nit: A few of the comments in this file (and possibly elsewhere?) are missing trailing ".".

File: components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt (hunk 10 above)
Comment (travis79) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt]: It's fairly common in mobile to have extensions and stuff at the end of the file, but I see your point.  Maybe it would be _best_ if I just put it in it's very own file?

Comment (travis79) [reply components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt]: Nevermind, I remember now why I did it in the same file...  due to visibility warnings/issues of how Kotlin treats `private` as rather `file private` than `class private`.  The worker needed access to call the updateExperiments() function.   I'll just move it to the top as you suggested.

File: components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsTest.kt (hunk 14 above)
Comment (travis79) [reply components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsTest.kt]: No, it turns out I couldn't get it to work with `verifyNoMoreInteractions` as there the interaction appeared to happening in a way to make this fail too.  I can test for the 1 interaction in resetExperiments that I know about, so I changed it to check for that and no more interactions.

File: components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsTest.kt (hunk 13 above)
Comment (travis79) [reply components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsTest.kt]: I cannot use `verifyNoMoreInteractions` here, this comment was relevant to the other place I had to `resetExperiments` minus the updater mocking and I just didn't update it to reflect the use-case.  I updated the comment to try and explain what I was seeing and why.
